// firebase-tools/lib/test/init/features/project.spec.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const chai_1 = require("chai");
const _ = require("lodash");
const sinon = require("sinon");
const project_1 = require("../../../init/features/project");
const projectManager = require("../../../management/projects");
const prompt = require("../../../prompt");
const TEST_FIREBASE_PROJECT = {
    projectId: "my-project-123",
    projectNumber: "123456789",
    displayName: "my-project",
    name: "projects/my-project",
    resources: {
        hostingSite: "my-project",
        realtimeDatabaseInstance: "my-project",
        storageBucket: "my-project.appspot.com",
        locationId: "us-central",
    },
};
const ANOTHER_FIREBASE_PROJECT = {
    projectId: "another-project",
    projectNumber: "987654321",
    displayName: "another-project",
    name: "projects/another-project",
    resources: {},
};
const TEST_PROJECT_INFO = {
    id: "my-project-123",
    label: "my-project-123 (my-project)",
    instance: "my-project",
    location: "us-central",
};
describe("project", () => {
    const sandbox = sinon.createSandbox();
    let getProjectStub;
    let createFirebaseProjectStub;
    let getOrPromptProjectStub;
    let addFirebaseProjectStub;
    let promptAvailableProjectIdStub;
    let promptOnceStub;
    let promptStub;
    beforeEach(() => {
        getProjectStub = sandbox.stub(projectManager, "getFirebaseProject");
        createFirebaseProjectStub = sandbox.stub(projectManager, "createFirebaseProjectAndLog");
        getOrPromptProjectStub = sandbox.stub(projectManager, "getOrPromptProject");
        addFirebaseProjectStub = sandbox.stub(projectManager, "addFirebaseToCloudProjectAndLog");
        promptAvailableProjectIdStub = sandbox.stub(projectManager, "promptAvailableProjectId");
        promptStub = sandbox.stub(prompt, "prompt").throws("Unexpected prompt call");
        promptOnceStub = sandbox.stub(prompt, "promptOnce").throws("Unexpected promptOnce call");
    });
    afterEach(() => {
        sandbox.restore();
    });
    describe("doSetup", () => {
        describe('with "Use an existing project" option', () => {
            it("should set up the correct properties in the project", () => __awaiter(this, void 0, void 0, function* () {
                const options = { project: "my-project" };
                const setup = { config: {}, rcfile: {} };
                promptOnceStub.onFirstCall().resolves("Use an existing project");
                getOrPromptProjectStub.onFirstCall().resolves(TEST_FIREBASE_PROJECT);
                yield project_1.doSetup(setup, {}, options);
                chai_1.expect(_.get(setup, "projectId")).to.deep.equal("my-project-123");
                chai_1.expect(_.get(setup, "instance")).to.deep.equal("my-project");
                chai_1.expect(_.get(setup, "projectLocation")).to.deep.equal("us-central");
                chai_1.expect(_.get(setup.rcfile, "projects.default")).to.deep.equal("my-project-123");
                chai_1.expect(promptOnceStub).to.be.calledOnce;
                chai_1.expect(getOrPromptProjectStub).to.be.calledOnceWith(options);
            }));
        });
        describe('with "Create a new project" option', () => {
            it("should create a new project and set up the correct properties", () => __awaiter(this, void 0, void 0, function* () {
                const options = {};
                const setup = { config: {}, rcfile: {} };
                promptOnceStub.onFirstCall().resolves("Create a new project");
                const fakePromptFn = (promptAnswer) => {
                    promptAnswer.projectId = "my-project-123";
                    promptAnswer.displayName = "my-project";
                };
                promptStub
                    .withArgs({}, projectManager.PROJECTS_CREATE_QUESTIONS)
                    .onFirstCall()
                    .callsFake(fakePromptFn);
                createFirebaseProjectStub.resolves(TEST_FIREBASE_PROJECT);
                yield project_1.doSetup(setup, {}, options);
                chai_1.expect(_.get(setup, "projectId")).to.deep.equal("my-project-123");
                chai_1.expect(_.get(setup, "instance")).to.deep.equal("my-project");
                chai_1.expect(_.get(setup, "projectLocation")).to.deep.equal("us-central");
                chai_1.expect(_.get(setup.rcfile, "projects.default")).to.deep.equal("my-project-123");
                chai_1.expect(promptOnceStub).to.be.calledOnce;
                chai_1.expect(promptStub).to.be.calledOnce;
                chai_1.expect(createFirebaseProjectStub).to.be.calledOnceWith("my-project-123", {
                    displayName: "my-project",
                });
            }));
            it("should throw if project ID is empty after prompt", () => __awaiter(this, void 0, void 0, function* () {
                const options = {};
                const setup = { config: {}, rcfile: {} };
                promptOnceStub.onFirstCall().resolves("Create a new project");
                const fakePromptFn = (promptAnswer) => {
                    promptAnswer.projectId = "";
                };
                promptStub
                    .withArgs({}, projectManager.PROJECTS_CREATE_QUESTIONS)
                    .onFirstCall()
                    .callsFake(fakePromptFn);
                let err;
                try {
                    yield project_1.doSetup(setup, {}, options);
                }
                catch (e) {
                    err = e;
                }
                chai_1.expect(err.message).to.equal("Project ID cannot be empty");
                chai_1.expect(promptOnceStub).to.be.calledOnce;
                chai_1.expect(promptStub).to.be.calledOnce;
                chai_1.expect(createFirebaseProjectStub).to.be.not.called;
            }));
        });
        describe('with "Add Firebase resources to GCP project" option', () => {
            it("should add firebase resources and set up the correct properties", () => __awaiter(this, void 0, void 0, function* () {
                const options = {};
                const setup = { config: {}, rcfile: {} };
                promptOnceStub
                    .onFirstCall()
                    .resolves("Add Firebase to an existing Google Cloud Platform project");
                promptAvailableProjectIdStub.onFirstCall().resolves("my-project-123");
                addFirebaseProjectStub.onFirstCall().resolves(TEST_FIREBASE_PROJECT);
                yield project_1.doSetup(setup, {}, options);
                chai_1.expect(_.get(setup, "projectId")).to.deep.equal("my-project-123");
                chai_1.expect(_.get(setup, "instance")).to.deep.equal("my-project");
                chai_1.expect(_.get(setup, "projectLocation")).to.deep.equal("us-central");
                chai_1.expect(_.get(setup.rcfile, "projects.default")).to.deep.equal("my-project-123");
                chai_1.expect(promptOnceStub).to.be.calledOnce;
                chai_1.expect(promptAvailableProjectIdStub).to.be.calledOnce;
                chai_1.expect(addFirebaseProjectStub).to.be.calledOnceWith("my-project-123");
            }));
            it("should throw if project ID is empty after prompt", () => __awaiter(this, void 0, void 0, function* () {
                const options = {};
                const setup = { config: {}, rcfile: {} };
                promptOnceStub
                    .onFirstCall()
                    .resolves("Add Firebase to an existing Google Cloud Platform project");
                promptAvailableProjectIdStub.onFirstCall().resolves("");
                let err;
                try {
                    yield project_1.doSetup(setup, {}, options);
                }
                catch (e) {
                    err = e;
                }
                chai_1.expect(err.message).to.equal("Project ID cannot be empty");
                chai_1.expect(promptOnceStub).to.be.calledOnce;
                chai_1.expect(promptAvailableProjectIdStub).to.be.calledOnce;
                chai_1.expect(addFirebaseProjectStub).to.be.not.called;
            }));
        });
        describe(`with "Don't set up a default project" option`, () => {
            it("should set up the correct properties when not choosing a project", () => __awaiter(this, void 0, void 0, function* () {
                const options = {};
                const setup = { config: {}, rcfile: {} };
                promptOnceStub.resolves("Don't set up a default project");
                yield project_1.doSetup(setup, {}, options);
                chai_1.expect(setup).to.deep.equal({ config: {}, rcfile: {}, project: {} });
                chai_1.expect(promptOnceStub).to.be.calledOnce;
            }));
        });
        describe("with defined .firebaserc file", () => {
            let options;
            let setup;
            beforeEach(() => {
                options = {};
                setup = { config: {}, rcfile: { projects: { default: "my-project-123" } } };
                getProjectStub.onFirstCall().resolves(TEST_FIREBASE_PROJECT);
            });
            it("should not prompt", () => __awaiter(this, void 0, void 0, function* () {
                yield project_1.doSetup(setup, {}, options);
                chai_1.expect(promptOnceStub).to.be.not.called;
                chai_1.expect(promptStub).to.be.not.called;
            }));
            it("should set project location even if .firebaserc is already set up", () => __awaiter(this, void 0, void 0, function* () {
                yield project_1.doSetup(setup, {}, options);
                chai_1.expect(_.get(setup, "projectId")).to.equal("my-project-123");
                chai_1.expect(_.get(setup, "projectLocation")).to.equal("us-central");
                chai_1.expect(getProjectStub).to.be.calledOnceWith("my-project-123");
            }));
        });
    });
});
